lib/ansible/modules/cloud/amazon/ec2_vpc_endpoint.py

```diff
@@ -0,0 +1,347 @@
+    status_achieved = False
+
+    for x in range(0, max_retries):
+        try:
+            resource = get_endpoints(client, module, resource_id)['VpcEndpoints'][0]
+            if resource['State'] == status:
+                status_achieved = True
+                break
+            else:
+                time.sleep(polling_increment_secs)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg=str(e))
+
+    return status_achieved, resource
+
+
+def get_endpoints(client, module, resource_id=None):
+    params = dict()
+    if resource_id:
+        params['VpcEndpointIds'] = [resource_id]
+
+    result = json.loads(json.dumps(client.describe_vpc_endpoints(**params), default=date_handler))
+    return result
+
+
+def setup_creation(client, module):
+    endpoint_found = False
+    if not module.params.get('vpc_id'):
+        module.fail_json(msg='vpc_id is a required paramater')
+    else:
+        vpc_id = module.params.get('vpc_id')
+    if not module.params.get('service'):
+        module.fail_json(msg='a valid service is a required paramater')
+    else:
+        service_name = module.params.get('service')
+
+    if module.params.get('route_table_ids'):
+        route_table_ids = module.params.get('route_table_ids')
+        existing_endpoints = get_endpoints(client, module)
+        for endpoint in existing_endpoints['VpcEndpoints']:
+            if endpoint['VpcId'] == vpc_id and endpoint['ServiceName'] == service_name:
+                sorted_endpoint_rt_ids = sorted(endpoint['RouteTableIds'])
+                sorted_route_table_ids = sorted(route_table_ids)
+                if cmp(sorted_endpoint_rt_ids, sorted_route_table_ids) == 0:
+                    changed = False
+                    endpoint_found = True
+                    result = endpoint
+
+    if not endpoint_found:
+        changed, result = create_vpc_endpoint(client, module)
+
+    return changed, json.loads(json.dumps(result, default=date_handler))
+
+
+def create_vpc_endpoint(client, module):
+    params = dict()
+    changed = False
+    token_provided = False
+    params['VpcId'] = module.params.get('vpc_id')
+    params['ServiceName'] = module.params.get('service')
+    params['DryRun'] = module.check_mode
+
+    if module.params.get('route_table_ids'):
+        params['RouteTableIds'] = module.params.get('route_table_ids')
+
+    if module.params.get('client_token'):
+        token_provided = True
+        request_time = datetime.datetime.utcnow()
+        params['ClientToken'] = module.params.get('client_token')
+
+    if module.params.get('policy'):
+        try:
+            policy = json.loads(module.params.get('policy'))
+        except ValueError as e:
+            try:
+                with open(module.params.get('policy'), 'r') as json_data:
+                    policy = json.load(json_data)
+                    json_data.close()
+            except Exception as e:
+                module.fail_json(msg=str(e))
```

Comment:
For exceptions, include `exception=traceback.format_exc()` so the full error is preserved.